src/jit/emitxarch.cpp

```diff
@@ -2901,174 +2901,282 @@ void emitter::emitInsStoreLcl(instruction ins, emitAttr attr, GenTreeLclVarCommo
 
-    // First handle the simple non-memory cases
-    //
-    if ((mem == nullptr) && (lclField == nullptr) && (lclVar == nullptr) && (tmpDsc == nullptr))
-    {
-        if (intConst != nullptr)
-        {
-            // reg, immed
-            assert(!dst->isContained());
+                    break;
+                }
 
-            emitIns_R_I(ins, attr, dst->gtRegNum, intConst->IconValue());
-            // TODO-XArch-Bug?: does the caller call regTracker.rsTrackRegTrash(dst->gtRegNum) or
-            // rsTrackRegIntCns(dst->gtRegNum, intConst->IconValue()) (as appropriate)?
-        }
-        else if (dblConst != nullptr)
-        {
-            // Emit a data section constant for float or double constant.
-            CORINFO_FIELD_HANDLE hnd = emitFltOrDblConst(dblConst->AsDblCon()->gtDconVal, emitTypeSize(dblConst));
+                case GT_CLS_VAR_ADDR:
+                {
+                    if (memOp == src)
+                    {
+                        assert(otrOp == dst);
+                        assert(cnsOp == nullptr);
 
-            emitIns_R_C(ins, attr, dst->gtRegNum, hnd, 0);
+                        if (instrHasImplicitRegPairDest(ins))
+                        {
+                            // src is a class static variable
+                            // dst is implicit - RDX:RAX
+                            emitIns_C(ins, attr, memBase->gtClsVar.gtClsVarHnd, 0);
+                        }
+                        else
+                        {
+                            // src is a class static variable
+                            // dst is a register
+                            emitIns_R_C(ins, attr, dst->gtRegNum, memBase->gtClsVar.gtClsVarHnd, 0);
+                        }
+                    }
+                    else
+                    {
+                        assert(memOp == dst);
+
+                        if (cnsOp != nullptr)
+                        {
+                            assert(cnsOp == src);
+                            assert(otrOp == nullptr);
+                            assert(src->IsCnsIntOrI());
+
+                            // src is an contained immediate
+                            // dst is a class static variable
+                            emitIns_C_I(ins, attr, memBase->gtClsVar.gtClsVarHnd, 0, (int)src->gtIntConCommon.IconValue());
+                        }
+                        else
+                        {
+                            assert(otrOp == src);
+
+                            // src is a register
+                            // dst is a class static variable
+                            emitIns_C_R(ins, attr, memBase->gtClsVar.gtClsVarHnd, src->gtRegNum, 0);
+                        }
+                    }
+
+                    return dst->gtRegNum;
+                }
+
+                default: // Addressing mode [regBase + regIndex * Scale + const]
+                {
+                    instrDesc* id = nullptr;
```

Comment:
I want to refactor this code separately (as part of https://github.com/dotnet/coreclr/issues/15828).

Reply:
After this is refactored, we could also rename the method to `genInsBinary` and move it to Codegen, which may be a more appropriate place.